api/documents/collection_ref.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2021 Nitric Technologies Pty Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package documents
+
+import (
+	"fmt"
+
+	v1 "github.com/nitrictech/go-sdk/interfaces/nitric/v1"
+)
+
+// Collection
+type CollectionRef interface {
+	Doc(string) DocumentRef
+	Query() Query
+	Parent() DocumentRef
+	toWire() *v1.Collection
+}
+
+type collectionRefImpl struct {
+	name           string
+	dc             v1.DocumentServiceClient
+	parentDocument DocumentRef
+}
+
+func (c *collectionRefImpl) Query() Query {
+	return newQuery(c, c.dc)
+}
+
+// Doc - Return a document reference for this collection
+func (c *collectionRefImpl) Doc(key string) DocumentRef {
+	return &documentRefImpl{
+		key: key,
+		dc:  c.dc,
+		col: c,
+	}
+}
+
+func (c *collectionRefImpl) Parent() DocumentRef {
+	return c.parentDocument
+}
+
+// toWire - tranlates a Collection for on-wire transport
+func (c *collectionRefImpl) toWire() *v1.Collection {
+	if c.parentDocument != nil {
+		return &v1.Collection{
+			Name:   c.name,
+			Parent: c.parentDocument.toWireKey(),
+		}
+	} else {
+		return &v1.Collection{
+			Name: c.name,
+		}
+	}
+}
+
+// converts a wire collection to a collection reference
+func collectionRefFromWire(dc v1.DocumentServiceClient, c *v1.Collection) (CollectionRef, error) {
+	if dc == nil {
+		return nil, fmt.Errorf("a document client is required for a collection reference")
+	}
+
+	if c.GetParent() == nil {
+		return &collectionRefImpl{
+			name: c.GetName(),
+			dc:   dc,
+		}, nil
+	} else {
+		pd, err := documentRefFromWireKey(dc, c.GetParent())
+		if err != nil {
+			return nil, err
```

Comment:
Might be useful to wrap this error to provide context, especially as the depth will be deeper in future.

Reply:
The error highlighted in this line isn't related to the collection depth, but I've crafted an explicit error for the collection depth exceed error.